internal/cloud/azure/application_identity.go

```diff
@@ -45,38 +46,53 @@ func (c *AzureConfig) NewServicePrincipalsClient(ctx context.Context) (ServicePr
 }
 
 type ApplicationIdentityConfig struct {
-	ID            string
-	Name          string
-	ApplicationID string
-	ClientID      string
-	ClientSecret  string
+	ID                     string
+	Name                   string
+	ApplicationID          string
+	ServicePrincipalID     string
+	ServicePrincipalSecret string
 }
 
 func CreateApplicationIdentity(ctx context.Context, config *ApplicationIdentityConfig, appClient ApplicationClient, spClient ServicePrincipalsClient) error {
-	app, _, appErr := appClient.Create(ctx, msgraph.Application{
-		DisplayName: &config.Name,
-	})
-	if appErr != nil {
-		return appErr
+	// The ID is the service principal ID, so technically theres a chance that last time the App was made
+	// but the SP was not. We need to check if the Application was made and skip creation if it was.
+	if config.ApplicationID == "" {
+		app, _, appErr := appClient.Create(ctx, msgraph.Application{
+			DisplayName: &config.Name,
+		})
+		if appErr != nil {
+			return appErr
+		}
+		config.ApplicationID = *app.ID
+	}
+
+	// fetch the application to make sure it exists
+	app, _, err := appClient.Get(ctx, config.ApplicationID, odata.Query{})
+	if err != nil {
+		return fmt.Errorf("error retrieving Application with object ID %v: %w", config.ApplicationID, err)
 	}
-	config.ApplicationID = *app.ID
-	config.ID = *app.ID
 
 	sp, _, spErr := spClient.Create(ctx, msgraph.ServicePrincipal{
 		AppId: app.AppId,
 	})
 	if spErr != nil {
 		return spErr
 	}
-	config.ClientID = *sp.ID
+	config.ServicePrincipalID = *sp.ID
+	config.ID = *sp.ID
 
-	spSecret, _, secretErr := spClient.AddPassword(ctx, *sp.ID, msgraph.PasswordCredential{
-		KeyId: sp.AppId,
-	})
+	// fetch the service principal to make sure it exists
+	_, _, spCheckErr := spClient.Get(ctx, config.ServicePrincipalID, odata.Query{})
+	if spCheckErr != nil {
+		return fmt.Errorf("error retrieving Service Principal with ID %v: %w", config.ServicePrincipalID, spCheckErr)
+	}
+
+	// Technically the secret is only needed for Kubernetes until the support workload identity. Maybe we make this on a conditional?
```

Comment:
Id prefer not to expose a cloud specific boolean, is there enough information to know its going to be running on kubernetes?

Reply:
> Id prefer not to expose a cloud specific boolean, is there enough information to know its going to be running on kubernetes?

Agreed on not exposing the boolean, which is why it's hardcoded to creation for now.  There currently isn't enough information to determine if its Kubernetes or not. I figure we should move a `kubernetes` block to the top level (@WillBeebe [added a `kubernetes` specific block to the GCP specific inputs](https://github.com/massdriver-cloud/terraform-provider-mdxc/blob/main/internal/provider/resource_application_identity.go#L50-L63).

Both GCP and Azure have this problem of needed to take different actions if its a kubernetes cluster, though Azure hopefully only until AKS workload identity is supported.  In AWS this is handled via the `assume_role_policy` which we are currently doing in terraform locals instead of the provider (though we could move it into the provider).

Reply:
Stretched the legs of this PR w/ Michael this afternoon. We ended up needing the service principal password to pull from the Azure container repo. 

I'd say we merge this, the terraform module, and the app template. Michael has the module and template in really good shape and they're almost ready for review. 

After that pass, we can do k8s mdx and I can add a `kubernetes` block here if we need it. LMK what you think.